boards/cats_rev1Pro/src/tasks/task_preprocessing.c

```diff
@@ -0,0 +1,240 @@
+      SI_data->gyro.z += (float32_t)global_imu[i].gyro_z * gyro_info[i].conversion_to_SI;
+    }
+  }
+
+#if NUM_ACCELEROMETER > 0
+  /* If all IMUs have been eliminated use high G accel */
+  if (counter == 0) {
+    for (int i = 0; i < NUM_ACCELEROMETER; i++) {
+      if (elimination_data->faulty_acc[i] == 0) {
+        counter++;
+        SI_data->accel.x += (float32_t)global_imu[i].acc_x * acc_info[NUM_IMU + i].conversion_to_SI;
+        SI_data->accel.y += (float32_t)global_imu[i].acc_y * acc_info[NUM_IMU + i].conversion_to_SI;
+        SI_data->accel.z += (float32_t)global_imu[i].acc_z * acc_info[NUM_IMU + i].conversion_to_SI;
+      }
+    }
+  }
+#endif
+
+  /* average for SI data */
+  if (counter > 0) {
+    SI_data->accel.x /= counter;
+    SI_data->accel.y /= counter;
+    SI_data->accel.z /= counter;
+    SI_data->gyro.x /= counter;
+    SI_data->gyro.y /= counter;
+    SI_data->gyro.z /= counter;
+    clear_error(CATS_ERR_FILTER_ACC);
+  } else {
+    SI_data->accel = SI_data_old->accel;
+    SI_data->gyro = SI_data_old->gyro;
+    add_error(CATS_ERR_FILTER_ACC);
+  }
+
+#endif
+
+#if NUM_BARO > 0
+  counter = 0;
+  SI_data->pressure.v = 0;
+  for (int i = 0; i < NUM_BARO; i++) {
+    if (elimination_data->faulty_baro[i] == 0) {
+      counter++;
+      SI_data->pressure.v += (float32_t)global_baro[i].pressure * baro_info[i].conversion_to_SI;
+    }
+  }
+  if (counter > 0) {
+    SI_data->pressure.v /= counter;
+    clear_error(CATS_ERR_FILTER_HEIGHT);
+  } else {
+    SI_data->pressure = SI_data_old->pressure;
+    add_error(CATS_ERR_FILTER_HEIGHT);
+  }
+#endif
+
+#if NUM_MAGNETO > 0
+  counter = 0;
+  SI_data->mag.x = 0;
+  SI_data->mag.y = 0;
+  SI_data->mag.z = 0;
+  for (int i = 0; i < NUM_MAGNETO; i++) {
+    if (elimination_data->faulty_mag[i] == 0) {
+      counter++;
+      SI_data->mag.x += (float32_t)global_magneto[i].magneto_x * mag_info[i].conversion_to_SI;
+      SI_data->mag.y += (float32_t)global_magneto[i].magneto_y * mag_info[i].conversion_to_SI;
+      SI_data->mag.z += (float32_t)global_magneto[i].magneto_z * mag_info[i].conversion_to_SI;
+    }
+  }
+  if (counter > 0) {
+    SI_data->mag.x /= counter;
+    SI_data->mag.y /= counter;
+    SI_data->mag.z /= counter;
+  } else {
+    /* Todo: Add error */
+    SI_data->mag = SI_data_old->mag;
+  }
+#endif
+}
+
+static void median_filter(median_filter_t *filter_data, state_estimation_input_t *state_data) {
+  /* Insert into array */
+  filter_data->acc[filter_data->counter] = state_data->acceleration_z;
+  filter_data->height_AGL[filter_data->counter] = state_data->height_AGL;
+
+  /* Update Counter */
+  filter_data->counter++;
+  filter_data->counter = filter_data->counter % MEDIAN_FILTER_SIZE;
+
+  /* Filter data */
+  state_data->acceleration_z = median(filter_data->acc);
+  state_data->height_AGL = median(filter_data->height_AGL);
+}
+
+static void transform_data(float32_t pressure_0, state_estimation_input_t *state_data, SI_data_t *SI_data,
```

Comment:
if *SI_data isn't modified in this function change the signature to
```
const SI_data_t * SI_data
```

Also check all other functions with pointers that you are introducing in this PR. It might give us some more optimization and the callers of these functions are safe in knowing which data will be modified and which not